[MRG+2] Make CD use fused types

yenchenlin opened the pull request:
According to #5464, current implementation of **ElasticNet** and **Lasso** in scikit-learn constrain the input to be `np.float64`, which is a waste of space.

This PR try to make CD algorithms support fused types when fitting `np.float32` dense data and therefore reduce redundant data copy.
- [x] Make inline helper functions support fused types
- [x] Make dense CD ElasticNet support fused types
- [x] Add warning when alpha close to zero and X is `np.float32`
- [x] Add tests

****UPDATE 7/7**

Here is the memory profiling results when fitting `np.float32` data:
- master

![master](https://cloud.githubusercontent.com/assets/7057863/16658531/4fae265c-4499-11e6-98c2-2905f6a72c41.png)
- this branch

![float32](https://cloud.githubusercontent.com/assets/7057863/16658248/41bc5858-4498-11e6-8050-c1fa1a0c9004.png)

****UPDATE 7/12**

Here is the memory profiling results when fitting **sparse** `np.float32` data:
- master

![64](https://cloud.githubusercontent.com/assets/7057863/16752698/416bed74-4816-11e6-8d55-513cbf62f597.png)
- this branch

![32](https://cloud.githubusercontent.com/assets/7057863/16752696/3e2af95c-4816-11e6-95ed-2a5154267a6f.png)

yenchenlin committed:
Make helper functions in cd use fused types

agramfort commented:
do you expect to merge just this or is it wip?

yenchenlin commented:
@agramfort Yeah I am thinking of merging just these as a start point.

jnothman commented:
Well, I think this is agreeable, but I'm not sure what value there is in merging it separately to something we can benchmark. For instance, you've fused `fmax` and `fsign`, but reused the `libc` `fabs` which explicitly operates on a double (as opposed to `fabsl`). Are you sure we benefit from fused implementations of max and sign?

So I think we want to review this as a whole.

yenchenlin committed:
Import cblas float functions

yenchenlin committed:
Make enet_coordinate_descent support fused types

yenchenlin commented:
Thanks @jnothman @agramfort .
Ah yeah you are right, will do 💪

yenchenlin commented:
Currently it is still not working.

Here is my test script:

```python
import numpy as np
from sklearn.linear_model.coordinate_descent import ElasticNet

X = np.array([[-1.], [0.], [1.]], dtype=np.float32)
Y = np.array([-1, 0, 1], dtype=np.float32)       # just a straight line
T = np.array([[2.], [3.], [4.]], dtype=np.float32)  # test sample

print X.dtype

clf = ElasticNet(alpha=1e-8, l1_ratio=1.0, precompute=False)
clf.fit(X, Y)
pred = clf.predict(T)

print clf.coef_.dtype
print pred.dtype
```

It will output that both `clf.coef` and `pred` are type `float64`.

yenchenlin committed:
Make dense case work

yenchenlin committed:
Refactor format

yenchenlin commented:
Here are the profiling results, I'm still looking for the main overhead ...
The execution time difference is huge ...

- master
![master](https://cloud.githubusercontent.com/assets/7057863/16541036/52ae9a22-40ab-11e6-8b12-2e647cbbb5d2.png)

- this PR
![cd-fused types](https://cloud.githubusercontent.com/assets/7057863/16541037/52ce88dc-40ab-11e6-888b-5c96d6dd594c.png)

yenchenlin committed:
Remove redundant change

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -175,108 +199,212 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
-
-                # np.dot(R.T, y)
-                gap += (alpha * l1_norm - const * ddot(
+        if floating is double:
```
Are you absolutely certain we can't do this with fused types? What prohibits it?

yenchenlin replied:
This algorithm uses lots of C pointer such as `<DOUBLE*>`, we can't do `<floating*>`.

jnothman replied:
You're saying Cython disallows fused type pointers, or typecasts? Is that incapability documented?

Could we use typecasts if we were working with typed memoryviews?

jnothman commented:
Could you use the line-based memory profiling to see where that sharp increase in memory consumption is coming in?

jnothman commented:
Sorry, that was silly; only appropriate if the bad memory usage is in Python code.

jnothman commented on sklearn/linear_model/base.py:
```diff
@@ -557,8 +557,8 @@ def _pre_fit(X, y, Xy, precompute, normalize, fit_intercept, copy):
-            fit_intercept and not np.allclose(X_offset, np.zeros(n_features)) or
-            normalize and not np.allclose(X_scale, np.ones(n_features))):
+            fit_intercept and not np.allclose(X_offset, np.zeros(n_features))
+            or normalize and not np.allclose(X_scale, np.ones(n_features))):
```
I don't see why this is an improvement, or why it's in this PR.

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -105,29 +106,44 @@ cdef extern from "cblas.h":
-                            int max_iter, double tol,
+def enet_coordinate_descent(np.ndarray[floating, ndim=1] w,
+                            floating alpha, floating beta,
+                            np.ndarray[floating, ndim=2, mode='fortran'] X,
```
I assume your benchmarking script is using fortran-contiguous data.

MechCoder replied:
The parent classes force the contiguity of X to be fortran-contiguous.

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -105,29 +106,44 @@ cdef extern from "cblas.h":
+def enet_coordinate_descent(np.ndarray[floating, ndim=1] w,
+                            floating alpha, floating beta,
+                            np.ndarray[floating, ndim=2, mode='fortran'] X,
+                            np.ndarray[floating, ndim=1, mode='c'] y,
```
I'm not sure that we need to support `y`, `alpha`, `beta`, `w` or `tol` with lower precision than `DOUBLE`. Or is consistent type necessary to use BLAS?

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -143,26 +159,34 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
-    cdef double l1_norm
+    cdef np.ndarray[floating, ndim=1] XtA
+
+    if floating is float:
```
below you use `if floating is double`. Please be consistent.

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -175,108 +199,212 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
+            # R = y - np.dot(X, w)
+            for i in range(n_samples):
+                R[i] = y[i] - ddot(n_features,
+                                <DOUBLE*>(X.data + i * sizeof(DOUBLE)),
```
I don't get why these lines were written like this in the first place. Surely defining `cdef DOUBLE *X_data = X.data` above, then doing `&X[i]` here is clearer. I presume that would work with `cdef floating *X_data = X.data` too.

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -175,108 +199,212 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
+                            1, <DOUBLE*>R.data, 1)
+
+                    # tmp = (X[:,ii]*R).sum()
+                    tmp = ddot(n_samples,
```
If these BLAS functions are the reason you use the if/else case, you could locally do `ddot if floating is double else sdot` or perhaps make a flag variable for if `floating is double` to make this briefer. Indeed it might be possible to declare a function pointer with `floating *` parameters above, set it to `dot = ddot if floating is double else sdot` and see if the C compiler optimises out the fact that the pointer is never changed.

jnothman commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -375,13 +375,23 @@ def enet_path(X, y, l1_ratio=0.5, eps=1e-3, n_alphas=100, alphas=None,
     # We expect X and y to be already float64 Fortran ordered when bypassing
```
Comment needs updating.

yenchenlin committed:
Add cblas files

yenchenlin committed:
Avoid redundant code

yenchenlin committed:
Remove redundant c files and import

yenchenlin committed:
Recover unnecessary change

yenchenlin committed:
Update comment

yenchenlin commented:
Hello @jnothman & @MechCoder , thanks alot-alot-alot-alot for your patience and comments.

I've updated the PR description (including **new** memory profiling results), code, and addressed comments you gave before.

The remaining to-do tasks in my opinion are also listed in the main description of this PR.

yenchenlin committed:
Make coef_ type consistent

yenchenlin committed:
Test float32 input

yenchenlin committed:
Add user warning when fitting float32 data with small alpha

yenchenlin commented:
I've added the tests and the user warning for potential non-convergence error when fitting `np.float32` data with small `alpha`.

yenchenlin committed:
Fix bug

yenchenlin committed:
Change variable to floating type

yenchenlin committed:
Make cd sparse support fused types

yenchenlin committed:
Make CD support fused types when data is sparse

yenchenlin commented:
I also complete the sparse case, should I include that into this PR?

agramfort commented on sklearn/linear_model/base.py:
```diff
@@ -273,6 +273,11 @@ def _set_intercept(self, X_offset, y_offset, X_scale):
         """Set the intercept_
         """
         if self.fit_intercept:
+            if isinstance(self.coef_, np.ndarray):
```
self.coef_ will always be an ndarray otherwis the line below  self.coef_ / X_scale would not work.

jnothman replied:
I'm confused. Why is this method changing?

yenchenlin replied:
It's because when fitting `np.float32` data, `X_offset`, `y_offset`, and `X_scale` are of `np.float64` if not explicitly setting their type here, which would in turn makes `enet` fitting `np.float32` data have `np.float64` `coef_` in the end.

MechCoder replied:
Could you clarify why `X_offset` etc are of dtype np.float64?

In any case, since `self.coef_` is always a numpy array, you can get the dtype without doing the if check, right?

yenchenlin replied:
hmm ...

Following error will show without this line:

```
  File "/Users/YenChen/Desktop/Python/scikit-learn/sklearn/linear_model/tests/test_least_angle.py", line 371, in test_multitarget
    estimator.fit(X, Y)
  File "/Users/YenChen/Desktop/Python/scikit-learn/sklearn/linear_model/least_angle.py", line 705, in fit
    self._set_intercept(X_offset, y_offset, X_scale)
  File "/Users/YenChen/Desktop/Python/scikit-learn/sklearn/linear_model/base.py", line 276, in _set_intercept
    dtype = self.coef_.dtype
AttributeError: 'list' object has no attribute 'dtype'
```

MechCoder replied:
This can be done away with, I believe. Look at `_preprocess_data` and see for places where it is possible that `X_offset` can be explicitly can be `float64`

yenchenlin replied:
Since error happened at `dtype = self.coef_.dtype`, with the message:

```
AttributeError: 'list' object has no attribute 'dtype'
```

I think `self.coef_` is not always a numpy array?

MechCoder replied:
Yes, but that is not the issue right now. For example, this works:

```python
a = [1, 2, 3]
b = np.array([1, 2, 3])
a / b
array([ 1.,  1.,  1.])
```

The issue is that `X_offset` is being cast to `np.float64` in `_preprocess_data`, which should not happen..

agramfort commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -372,16 +372,26 @@ def enet_path(X, y, l1_ratio=0.5, eps=1e-3, n_alphas=100, alphas=None,
+            if Xy is not None:
+                # Xy should be a 1d contiguous array or a 2D C ordered array
+                Xy = check_array(Xy, dtype=np.float64, order='C', copy=False,
+                                 ensure_2d=False)
```
this is a big code dupe. Just introduce a dtype variable

jnothman replied:
please address this! and yet doesn't `check_array` allow for a list of dtypes in order of preference, so that `dtype=[np.float64, np.float32]` should do the trick?

jnothman replied:
oh of course, the arrays all need to be consistent, so that wouldn't do the trick.

jnothman replied:
still, using that for `X` then using `dtype=X.dtype` should work?

yenchenlin replied:
Yes thanks!

While technically, it;s `dtype=X.dtype.type`.
Again, maybe `check_array` should support `dtype=X.dtype`?

agramfort commented:
PR is not huge so yes add the sparse case

MechCoder commented:
@yenchenlin The travis failure is not spurious. You have references to `ATL_sasum` and `ATL_saxpy` which are defined in `"atlas_refalias1.h"` as `ATL_srefasum` and `ATL_srefaxpy`. You need to add `ATL_srefasum.h` and `ATL_srefaxpy.h`. You can get them from here https://searchcode.com/codesearch/view/86268628/ and https://searchcode.com/codesearch/view/86268548/. I added them and rebuild the package and it works

MechCoder commented:
Please add those and ping back. Thanks!

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -13,10 +13,16 @@ import numpy.linalg as linalg
 
 ctypedef np.float64_t DOUBLE
 ctypedef np.uint32_t UINT32_t
+ctypedef floating (*DOT)(int N, floating *X, int incX, floating *Y,
```
Nice; looks like you're getting the hang of this!

MechCoder replied:
neat

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -143,26 +154,47 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
+    cdef AXPY axpy
+    cdef ASUM asum
+
+    if floating is float:
```
I assume you get the idea that Cython compiles this as if there is no `if` here; the only if is when detecting the type passed in from Python. Then we hope that the C optimiser recognises these as variables that are constant locally, being assigned values that are constant globally (at linking time) and hence compiles this all to be identical to the former version...

jnothman commented:
In case you haven't you should also run a benchmark with 64-bit data, to make sure nothing has deteriorated.

jnothman commented:
Did you forget to commit your changes to the `pyx`?

yenchenlin commented:
Yeah you are right, I miss a commit when I'm cherry-picking 😄 
Also the sparse profiling results updated in the main description!

jnothman commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -664,3 +664,24 @@ def test_lasso_non_float_y():
+
+
+def test_enet_float_precision():
+	# Generate dataset
```
indentation

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -307,30 +338,54 @@ def sparse_enet_coordinate_descent(double[:] w,
+    cdef ASUM asum
+
+    if floating is float:
+        norm_cols_X = np.zeros(n_features, dtype=np.float32)
```
please use a dtype variable to avoid duplicating this.

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -307,30 +338,54 @@ def sparse_enet_coordinate_descent(double[:] w,
+    if floating is float:
+        norm_cols_X = np.zeros(n_features, dtype=np.float32)
+        n_tasks = y.strides[0] / sizeof(float)
+        X_T_R = np.zeros(n_features, dtype=np.float32)
```
and here etc.

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -13,10 +13,16 @@ import numpy.linalg as linalg
+ctypedef floating (*DOT)(int N, floating *X, int incX, floating *Y,
+                         int incY) nogil
+ctypedef void (*AXPY)(int N, floating alpha, floating *X, int incX,
+                          floating *Y, int incY) nogil
```
Indentation

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -13,10 +13,16 @@ import numpy.linalg as linalg
 ctypedef np.float64_t DOUBLE
 ctypedef np.uint32_t UINT32_t
+ctypedef floating (*DOT)(int N, floating *X, int incX, floating *Y,
+                         int incY) nogil
```
indentation ([PEP8](https://www.python.org/dev/peps/pep-0008/#indentation))

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -105,9 +111,14 @@ cdef extern from "cblas.h":
     double ddot "cblas_ddot"(int N, double *X, int incX, double *Y, int incY
                              ) nogil
+    float sdot "cblas_sdot"(int N, float *X, int incX, float *Y, int incY
+                             ) nogil
```
indentation

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -363,7 +418,7 @@ def sparse_enet_coordinate_descent(double[:] w,
 
         # tol *= np.dot(y, y)
-        tol *= ddot(n_samples, <DOUBLE*>&y[0], 1, <DOUBLE*>&y[0], 1)
+        tol *= dot(n_samples, <floating*>&y[0], 1, <floating*>&y[0], 1)
```
It's not clear to me that these typecasts are still necessary. Have you tried removing them?

yenchenlin replied:
Yes, but it will appear `Cannot assign type 'char *' to 'float *'`.
So typecasts are still necessary.

yenchenlin replied:
I'm wrong, it can be fixed by casting when declaring variables.

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -201,26 +233,26 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
-                           <DOUBLE*>(X.data + ii * n_samples * sizeof(DOUBLE)),
-                           1, <DOUBLE*>R.data, 1)
+                tmp = dot(n_samples,
+                        <floating*>(X.data + ii * n_samples * sizeof(floating)),
```
Indentation

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -201,26 +233,26 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
-                          <DOUBLE*>(X.data + ii * n_samples * sizeof(DOUBLE)),
-                          1, <DOUBLE*>R.data, 1)
+                    axpy(n_samples, w_ii,
+                        <floating*>(X.data + ii * n_samples * sizeof(floating)),
```
Indentation

jnothman replied:
Could you try defining `X_data` and `y_data` above and then just using `&X_data[ii * n_samples]`??

jnothman replied:
I know that's not really your code, but all of these typecasts seem messy.

yenchenlin committed:
Add referenced src files

yenchenlin committed:
Avoid duplicated code

yenchenlin committed:
Avoid type casting

yenchenlin committed:
Fix indentation in test

yenchenlin committed:
Avoid type casting in sparse implementation

yenchenlin committed:
Fix indentation

yenchenlin committed:
Fix duplicated intialization code

yenchenlin committed:
Follow PEP8

yenchenlin commented:
Hello @MechCoder and @jnothman , I've addressed most of the comments here excluding benchmark `np.float64` data. I planned to do that after the CI is happy.

Do you have any idea why `PYTHON_ARCH=32` machine are complaining?

jnothman commented:
If necessary, you can write debug statements to stderr and ask AppVeyor to re-test, in order to at least trace which line the error is occurring at.

yenchenlin commented:
Thanks @jnothman .
I do find which line cause the error by comment out test code of test and repush this branch to AppVeyor. It's [this line](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/tests/test_coordinate_descent.py#L633)

I am trying methods provided from mailing list to rebuild 32bit PYTHON_ARCH env to get more information for debugging.

yenchenlin committed:
Raise tmp precision to double

yenchenlin commented:
Hello @MechCoder and @jnothman , sorry for interrupt.

Is there any difference to run [saxpy](https://github.com/yenchenlin/scikit-learn/blob/cd-fused-types/sklearn/linear_model/cd_fast.pyx#L243-L244) on 32-bit and 64-bit computer when fitting `np.float32` data?

Currently I'm using remote desktop to access appveyor's build worker in order to reproduce the error. Although the test failed on Windows without any error message, I use the naive `printf` debugging strategy to find out which line it crashes.

It fails at [this line](https://github.com/yenchenlin/scikit-learn/blob/cd-fused-types/sklearn/src/cblas/ATL_srefaxpy.c#L131) of `ATL_srefaxpy.c` during the second iteration of coordinate descent algorithm because [this condition](https://github.com/yenchenlin/scikit-learn/blob/cd-fused-types/sklearn/linear_model/cd_fast.pyx#L241) prevent `saxpy` to be called during the first iteration of the CD algorithm.

I know you are busy, but I've been stuck for two days and I really want to get this PR done!

jnothman commented:
Hmm... I assume the ATLAS code should be platform independent. If I were you I'd carefully check that the input was perfectly valid.

yenchenlin committed:
Add 64 bit computer check

yenchenlin committed:
Fix test

yenchenlin committed:
Add constraint

yenchenlin commented:
Hello @jnothman and @MechCoder ,
after a long reproducing-CI-environment marathon I found that this PR requires Numpy version >= 1.10.x and 64-bit computer to not suffer from precision issue.

And the main cause of precision problem is [this line](https://github.com/scikit-learn/scikit-learn/pull/6913/files#diff-5d8b82d506a543bd684a1d403d5941e0R260), which can cause precision problem after second iteration of the CD algorithm.

I've tried to use `dsdot` to raise the precision of `tmp` to double, but it only helps a little.

I also tried to raise `R`'s precision to `double`, and since I'm being constrained by cblas functions' arguments type, this is not doable currently.

yenchenlin commented:
Base on the above reason, I currently use a `bool` [here](https://github.com/scikit-learn/scikit-learn/pull/6913/files#diff-a84a355d447d21d71c4f82afbeba65edR32) to pass the test.

jnothman commented:
When you say precision issues, from which numpy < 1.10 and 32-bit architectures suffer, are you still talking about issues related to small alpha? Are they again failure-to-identify-convergence issues, or numerical precision of the optimised parameters?

yenchenlin commented:
No, it's not related to the alpha.

It's [this line](https://github.com/scikit-learn/scikit-learn/pull/6913/files#diff-5d8b82d506a543bd684a1d403d5941e0R260) which will cause numerical precision of the optimised parameters.

For example, elements' value of `R` may be 0.5 in the correct case, but its' value become 1.02e+18 after `saxpy` when suffering from precision issues.

May I ask how is the test run on your computer locally?

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -231,31 +269,27 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
             if (w_max == 0.0
-                    or d_w_max / w_max < d_w_tol
-                    or n_iter == max_iter - 1):
+                or d_w_max / w_max < d_w_tol
```
This is not PEP8 indentation.

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -201,26 +240,25 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
                 else:
                     w[ii] = (fsign(tmp) * fmax(fabs(tmp) - alpha, 0)
-                             / (norm_cols_X[ii] + beta))
+                            / (norm_cols_X[ii] + beta))
```
PEP8 indentation

yenchenlin replied:
I've fixed this, but it seems that some doctest is now failing.
Should I rebase master and force push the branch?

jnothman replied:
Doctest is someone else's problem:
https://github.com/scikit-learn/scikit-learn/pull/7083/files#r72380271

On 27 July 2016 at 14:40, Yen <notifications@github.com> wrote:

> In sklearn/linear_model/cd_fast.pyx
> <https://github.com/scikit-learn/scikit-learn/pull/6913#discussion_r72380018>
> :
>
> >
> >                  if positive and tmp < 0:
> >                      w[ii] = 0.0
> >                  else:
> >                      w[ii] = (fsign(tmp) * fmax(fabs(tmp) - alpha, 0)
> > -                             / (norm_cols_X[ii] + beta))
> > +                            / (norm_cols_X[ii] + beta))
>
> I've fixed this, but it seems that some doctest is now failing.
> Should I rebase master and force push the branch?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6913/files/dfaedcfa6374890580479e0cc9f927806af95179#r72380018>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz68CUpaNnaAxcmVP1BPghl8YwH4egks5qZuFPgaJpZM4I6SId>
> .
>

yenchenlin replied:
Okay!

jnothman commented:
> May I ask how is the test run on your computer locally?

Do you mean in my usual or a specific configuration?

yenchenlin commented:
> Do you mean in my usual or a specific configuration?

Sorry, I am silly; I'll make a simple example which can reproduce the issue and print the value of some important variables.

yenchenlin committed:
PEP 8

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -201,26 +240,25 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
-                           <DOUBLE*>(X.data + ii * n_samples * sizeof(DOUBLE)),
-                           1, <DOUBLE*>R.data, 1)
+                if floating is float:
+                    tmp = dsdot(n_samples, &X_data[ii * n_samples], 1, R_data, 1)
```
So does this help?

jnothman replied:
I suppose you wouldn't conclusively be able to tell, but you've clearly chosen not to *always* use `dsdot` instead of `sdot`.

yenchenlin replied:
Yeah it's because `tmp` won't be used to perform operation with other variables.

Actually, computation of `XtA` [here](https://github.com/scikit-learn/scikit-learn/pull/6913/files#diff-5d8b82d506a543bd684a1d403d5941e0R280) can use `dsdot` too.

jnothman replied:
oh of course, I was thinking of `sdsdot`

yenchenlin replied:
But if we use `dsdot` in [this line](https://github.com/scikit-learn/scikit-learn/pull/6913/files#diff-5d8b82d506a543bd684a1d403d5941e0R222), we may need to make both `R` and `y` become double, which will cause problems since `R` will be used to perform cblas operation with `X`.

jnothman replied:
Yes, I've realise. It's fine as is. Sorry I haven't got around to giving this a full review yet, but I'm off to sleep.

MechCoder replied:
Do we still need this?

yenchenlin replied:
I'll try to remove this and see if CI passed.

MechCoder commented:
Extremely sorry for the delays. It kind of worries me that I have to have a version of NumPy >= than 1.10 to avail this feature. 

Do you have a code snippet that produces completely different results in a version of NumPy lesser than 1.10 (and on a 64 bit computer) ? I find it surprising that upgrading NumPy can solve precision issues.

MechCoder commented:
Nvm, I can reproduce the test failure locally with Python 3.4 and NumPy 1.10.

jnothman commented:
> Nvm, I can reproduce the test failure locally with Python 3.4 and NumPy 1.10.
 Hide all checks

Test fails with numpy 1.10?

MechCoder commented:
Yes,
![screenshot from 2016-07-28 22 13 19](https://cloud.githubusercontent.com/assets/1867024/17238444/8a871072-5510-11e6-89b6-5c7fd9a8df3e.png)

yenchenlin commented:
Hello @MechCoder , it must be my carelessness.

May you proovide more environment detail?

My judgement for numpy version comes from I reinstall 5 Unix 64bit computers with numpy 1.10 in my school, and all of them works well. On the other hand, CI envs that passed the tests all have numpy >= 1.10, and that failed only have numpy version <= 1.10.

jnothman commented:
> all have numpy >= 1.10, and that failed only have numpy version <= 1.10.

Should one of those be `>` or `<` rather than `>=` and `<=`??

jnothman commented:
@MechCoder so as far as you're concerned, the criteria of np1.10 and 64bit arch and big alpha are not sufficient to avoid bad convergence issues?

MechCoder commented:
Yes, that is precisely my concern. I use `1.10.1` with a 64 bit version.

Could it also be because of how NumPy is built rather than the version number? (I'll verify this in a while)

yenchenlin commented:
@jnothman yeah, you are right.

@MechCoder thanks for your help!

MechCoder commented:
I cannot reproduce this on my work laptop with the oldest supported NumPy(1.7.2), SciPy (0.12.1) with Python3 on a 64-bit version. The Travis failures fail for Python2, with older versions of NumPy and SciPy. Could you check if you are able to reproduce the failures on a 64bit system with Python2 and the latest versions of NumPy / SciPy?

(This does not answer the previous question on why the tests fail for Python3 on my home computer, for which I would have to go home)

MechCoder commented:
Also, were tests passing locally before you addressed this comment (https://github.com/scikit-learn/scikit-learn/pull/6913#issuecomment-231558944) ?

MechCoder commented:
Before any further speculation, can you push a commit that adds a build here (https://github.com/scikit-learn/scikit-learn/blob/master/.travis.yml#L42) with `install_mkl` set to False and report the results?

yenchenlin commented:
Hello @MechCoder , yes the tests pass locally on my computer (and I use Python 2.7).

My mac and 5 Ubuntu computers I can find all run Python 2.7 with Numpy = 1.10 & Scipy = 0.17, and all of them can't reproduce the problem.

yenchenlin commented:
Also I followed your instruction and pushed  a commit [here](https://travis-ci.org/yenchenlin/scikit-learn).

MechCoder commented:
All right, it is just as I suspected. There is some problem with the `cblas` libraries you have added. In my laptop at home, I do not have atlas installed and so it depends on the cblas libraries added locally.
In short, this (https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/setup.py#L69) and (https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/setup.py#L23) this returns False.

I can reproduce it on my work laptop by this commit (https://github.com/MechCoder/scikit-learn/commit/df8ce54f402f254c0f8e0b76d94e1a746fc5953d) forcing it to use the cblas in sklearn/src/cblas

Can you double check the blas files you have added?

yenchenlin commented:
Hello @MechCoder , I can only reproduce the issue using your commit on my friend's old mac but not on my own mac.

Does the issue disappear when you remove the latest commit you added?

MechCoder commented on sklearn/src/cblas/cblas_dsdot.c:
```diff
@@ -0,0 +1,53 @@
+#include "atlas_level1.h"
+#include "cblas.h"
+
+double cblas_dsdot(const int N, const float *X, const int incX,
```
What is the difference between this and `cblas_sdot.c` that we have already have?

yenchenlin replied:
It takes `float` input arguments and output `double`.

`cblas_sdot.c` takes `double` input arguments and output `double`.

jnothman replied:
I think you must mean `float` and `float`

MechCoder commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -218,9 +257,8 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
-                    daxpy(n_samples, -w[ii],
-                          <DOUBLE*>(X.data + ii * n_samples * sizeof(DOUBLE)),
-                          1, <DOUBLE*>R.data, 1)
+                    axpy(n_samples, -w[ii], &X_data[ii * n_samples], 1,
```
It's the line that is creating a huge blowup even `maxiter` is set to 1 and `fiter` is the first feature. The input to this call are almost the same for both the float32 and float64 cases. Anything that is obviously wrong?

MechCoder replied:
I removed just this line and replaced it with

```python
R -= w[ii] * X[:, ii]
```

and it works as it is supposed to :/

jnothman replied:
Grr... and somehow what's broken is the newly included `ATL_srefaxpy`? It looks entirely innocuous, i.e. it's doing the same except multiplying by the weight for each element in turn.

I'm curious whether the bug appears if you reduce the compiler optimisation level.

fabianp committed:
Make saxpy have the same structure as daxpy

Hopefully this fixes the problems outlined in PR #6913

jnothman commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -27,6 +27,13 @@
 from . import cd_fast
+import struct
+
+if (struct.calcsize("P") * 8 == 64
```
Is this still all relevant?

yenchenlin replied:
Nope! I deleted them.

yenchenlin committed:
Remove wrong hardware test

yenchenlin committed:
Remove dsdot

MechCoder commented:
Should we fix `ElasticNetCV` and `LassoCV` as well in this PR? It should be pretty straightforward to extend.

MechCoder commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -664,3 +664,22 @@ def test_lasso_non_float_y():
+    X, y, X_test, y_test = build_dataset(n_samples=20, n_features=10)
+    # Here we have a small number of iterations, and thus the
+    # ElasticNet might not converge. This is to speed up tests
+    clf = ElasticNet(alpha=0.5, max_iter=100, precompute=False)
```
Sanity check for Lasso as well?

MechCoder replied:
Also check for `fit_intercept=True | False`

MechCoder commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -670,13 +677,14 @@ def fit(self, X, y, check_input=True):
-                             copy=self.copy_X and self.fit_intercept,
-                             multi_output=True, y_numeric=True)
-            y = check_array(y, dtype=np.float64, order='F', copy=False,
+            y = np.asarray(y)
```
Can this line be removed?

MechCoder commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -472,6 +475,10 @@ def enet_path(X, y, l1_ratio=0.5, eps=1e-3, n_alphas=100, alphas=None,
                           ' You might want' +
                           ' to increase the number of iterations',
                           ConvergenceWarning)
+            if X.dtype == np.float32:
```
Is this statement still valid? In any case, you should also check if `alpha` is lesser than a certain `tol`, so that your statement is valid.

yenchenlin replied:
Okay, value lower than 1e-8 tends to cause non-convergence issue.

1e-7 works well.

MechCoder commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -265,33 +294,30 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
-                            double tol, object rng, bint random=0,
+                            np.ndarray[floating, ndim=1] y,
+                            floating[:] X_mean, int max_iter,
+                            floating tol, object rng, bint random=0,
```
nitpick: I suppose we can live without `tol` being floating?

yenchenlin replied:
Yes, but then we may need to make `w_max` and other related variables become non-floating according to [this line](https://github.com/scikit-learn/scikit-learn/pull/6913/files#diff-5d8b82d506a543bd684a1d403d5941e0R267). (**Note**: `d_w_tol` = `tol`)

And since `w_max` is used to compare with `w` (which must be `floating`) at [this line](https://github.com/scikit-learn/scikit-learn/pull/6913/files#diff-5d8b82d506a543bd684a1d403d5941e0R263), things will getting too complicated in my opinion.

yenchenlin replied:
Also I tried to use `dsdot` to compute some other variable before and that showed really little benefit in reducing precision error.

MechCoder commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -105,29 +111,35 @@ cdef extern from "cblas.h":
+                            floating alpha, floating beta,
+                            np.ndarray[floating, ndim=2, mode='fortran'] X,
+                            np.ndarray[floating, ndim=1, mode='c'] y,
+                            int max_iter, floating tol,
```
same comment about `tol`

yenchenlin replied:
same answer [here](https://github.com/scikit-learn/scikit-learn/pull/6913/files#r74798141).

MechCoder replied:
yeah fine :p

MechCoder commented:
For the `ElasticNetCV` (and related classes), the precompute is set by default to "auto", which means for benefits in memory when `n_samples > n_features`, you need to make `enet_coordinate_descent_gram` handle float32 dtypes as well.

As for `ElasticNet` and related classes, we will be setting `precompute=False"` by default from this version, which means unless the user explicitly provides a precomputed matrix or sets it to True, we will get memory benefits.

I am fine with supporting fused types for `enet_coordinate_descent_gram` in a separate PR, but I would appreciate your and @jnothman ;s opinion as well.

jnothman commented:
I'm happy for this to be merged as soon as it is functional and of acceptable quality within its current scope, as long as low-hanging fruit for extension is at least ticketed with a new issue.

yenchenlin committed:
Remove redundant asarray

yenchenlin committed:
Add test for fit_intercept

MechCoder commented:
Yeah, that's fine.

yenchenlin committed:
Make _preprocess_data support other dtypes

yenchenlin commented:
@MechCoder you are right, it works well after making `_preprocess_data ` support not only `np.float64`.

jnothman commented:
> @MechCoder you are right, it works well after making `_preprocess_data` support not only np.float64.

I'm a little concerned there's not enough float32 testing of other linear models to know whether this change has broken anything.

yenchenlin committed:
Add concrete value

yenchenlin commented:
@jnothman , I think other linear models will call `check_X_y` to ensure data is `np.float64` before calling `_preprocess_data` if they don't support `np.float32`.

For example:

- [ridge.py](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/ridge.py#L941)
- [bayes.py](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/bayes.py#L363)

jnothman commented:
If you can confirm that for all cases, it's not exactly a rigorous test, but we're better safe than sorry.

yenchenlin commented:
I do a `fgrep` to find where `_preprocess_data` is called, here's the results:
```
./base.py:144:def [01;31m[K_preprocess_data[m[K(X, y, fit_intercept, normalize=False, copy=True,
./base.py:209:# TODO: _rescale_data should be factored into [01;31m[K_preprocess_data[m[K.
./base.py:270:    [01;31m[K_preprocess_data[m[K = staticmethod([01;31m[K_preprocess_data[m[K)
./base.py:515:        X, y, X_offset, y_offset, X_scale = self.[01;31m[K_preprocess_data[m[K(
./base.py:552:        X, y, X_offset, y_offset, X_scale = [01;31m[K_preprocess_data[m[K(
./base.py:557:        X, y, X_offset, y_offset, X_scale = [01;31m[K_preprocess_data[m[K(
./bayes.py:145:        X, y, X_offset, y_offset, X_scale = self.[01;31m[K_preprocess_data[m[K(
./bayes.py:368:        X, y, X_offset, y_offset, X_scale = self.[01;31m[K_preprocess_data[m[K(
./coordinate_descent.py:17:from .base import [01;31m[K_preprocess_data[m[K
./coordinate_descent.py:90:            X, y, _, _, _ = [01;31m[K_preprocess_data[m[K(X, y, fit_intercept,
./coordinate_descent.py:97:            _, _, X_offset, _, X_scale = [01;31m[K_preprocess_data[m[K(X, y, fit_intercept,
./coordinate_descent.py:1715:        X, y, X_offset, y_offset, X_scale = [01;31m[K_preprocess_data[m[K(
./least_angle.py:638:        X, y, X_offset, y_offset, X_scale = self.[01;31m[K_preprocess_data[m[K(X, y,
./least_angle.py:1447:        X, y, Xmean, ymean, Xstd = LinearModel.[01;31m[K_preprocess_data[m[K(
./randomized_l1.py:18:from .base import [01;31m[K_preprocess_data[m[K
./randomized_l1.py:74:    [01;31m[K_preprocess_data[m[K = staticmethod([01;31m[K_preprocess_data[m[K)
./randomized_l1.py:98:            self.[01;31m[K_preprocess_data[m[K(X, y, self.fit_intercept, self.normalize)
./randomized_l1.py:526:    def [01;31m[K_preprocess_data[m[K(self, X, y, fit_intercept, normalize=False):
./randomized_l1.py:528:        X, _, X_offset, _, X_scale = [01;31m[K_preprocess_data[m[K(X, y, fit_intercept,
./ridge.py:471:        X, y, X_offset, y_offset, X_scale = self.[01;31m[K_preprocess_data[m[K(
./ridge.py:946:        X, y, X_offset, y_offset, X_scale = LinearModel.[01;31m[K_preprocess_data[m[K(
```

For `bayes.py`, 
- [bayes.py:145]: [bayes.py:144](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/bayes.py#L144).
- [bayes.py:368]: [bayes.py:363](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/bayes.py#L363)

For `ridge.py`, 
- [ridge.py:471]: [ridge.py:464](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/ridge.py#L464).
- [ridge.py:946]: [ridge.py:946](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/ridge.py#L941).

It seems following classes in following files need to be tested:

- `base.py`
- `randomized_l1.py`
- `least_angle.py`

jnothman commented:
> Should I open another PR and make sure that X, y are of type np.float64 before _preprocess_data is called in the remaining files?

* base.py
* randomized_l1.py
* least_angle.py

Do you mean that you have checked and it is not the case?

jnothman commented:
no you can't have checked yet: base.py uses it in _pre_fit, which is used in OMP and CD. CD, at least, constrains to float64; OMP does not appear to.

yenchenlin commented:
Oh yeah ... sorry!

I think models which are affected by the changement of `base.py` are beyond the scope of this PR.
Maybe I should just typecast enet's `self.coef_` from `np.float64` to `np.float32` [here](https://github.com/scikit-learn/scikit-learn/pull/6913/files#diff-a84a355d447d21d71c4f82afbeba65edR739) when input X is of type `np.float32`?

yenchenlin commented:
Because the reason why I modified `base.py` is that I want to make sure `self.coef_`is of type `np.float32` when input X is `np.float32`.

yenchenlin committed:
Workaround

yenchenlin commented:
@jnothman @MechCoder 

Now I leave `base.py` as unchanged and typecast `self.coef_` of enet to type of `X`.
Doing this way, we don't need to worry about changing `base.py` would broke anything.

BTW, This seems to be the only remaining issue right now!

jnothman commented:
Yes, I agree we're close to merge here. I just need some time to look it over in full.

jnothman commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -664,3 +664,24 @@ def test_lasso_non_float_y():
+            coef[dtype] = clf.coef_
+
+        assert_array_almost_equal(coef[np.float32], coef[np.float64],
+                                  decimal=4)
```
should be checking `intercept_` too. and the dtypes of the `coef_` if appropriate.

jnothman commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -472,6 +475,11 @@ def enet_path(X, y, l1_ratio=0.5, eps=1e-3, n_alphas=100, alphas=None,
                           ' to increase the number of iterations',
                           ConvergenceWarning)
+            if X.dtype == np.float32:
+                warnings.warn('It may cause by precision issues' +
```
Reorder -> "Fitting float32 data with small ... may cause precision issues. Try ..."

jnothman replied:
I suspect this message should be part of the `ConvergenceWarning` above.

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -463,13 +513,13 @@ def sparse_enet_coordinate_descent(double[:] w,
-                l1_norm = dasum(n_features, <DOUBLE*>&w[0], 1)
+                l1_norm = asum(n_features, &w[0], 1)
 
                 # The expression inside ddot is equivalent to np.dot(R.T, y)
```
reference to ddot no longer applicable.

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -143,47 +155,72 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
 
     # initial value of the residuals
-    cdef np.ndarray[DOUBLE, ndim=1] R = np.empty(n_samples)
+    cdef np.ndarray[floating, ndim=1] R
```
Is there a reason not to initialise this here?

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -143,47 +155,72 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
+        axpy = daxpy
+        asum = dasum
+
+    R = np.empty(n_samples, dtype=dtype)
```
i.e. move this up?

jnothman commented on sklearn/linear_model/cd_fast.pyx:
```diff
@@ -143,47 +155,72 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
-    cdef double l1_norm
+    cdef np.ndarray[floating, ndim=1] XtA
+
+    # fused types version of BLAS functions
```
I find the placement of this a bit awkward. Could you please move it to before all the local variable definitions or after them?

jnothman commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -729,6 +737,9 @@ def fit(self, X, y, check_input=True):
         self._set_intercept(X_offset, y_offset, X_scale)
 
+        # workaround since _set_intercept will cast self.coef_ into float64
+        self.coef_ = np.asarray(self.coef_, dtype=X.dtype)
```
do we need to do the same for `intercept_`

jnothman replied:
I think we don't need to.

I'm not even certain we need to make `coef_` match the dtype.

yenchenlin replied:
In my opinion, it would be confusing if I fit `np.float32` training data, and get prediction result with`np.float64`.

So I think we need to make `coef_` match the type.

yenchenlin replied:
Also `intercept_`

MechCoder replied:
So:

```python
from sklearn.linear_model.base import _preprocess_data
X = rng.randn(5, 5).astype(np.float32)
y = rng.randn(5).astype(np.float32)
for fit_intercept in [True, False]:
    X, y, X_offset, y_offset, X_scale = _preprocess_data(X, y, fit_intercept=True)
    X_scale.dtype
    dtype('float64')
```

I suppose we can do away with this "workaround" if `X_scale` is set to the right dtype in `_preprocess_data`, i.e replace `X_scale = np.ones(X.shape[1])` with `X_scale = np.ones(X.shape[1], dtype=X.dtype)` etc wherever applicable.

yenchenlin replied:
But will this affect other models?

It would be safer to not touch `_preprocess_data` since it is called by lots of other models.
See [this comment](https://github.com/scikit-learn/scikit-learn/pull/6913#issuecomment-240020536).

MechCoder replied:
Sorry, I do not get it. Your comment suggests that the input data is explicitly cast to float64 before the call to `_preprocess_data`. So what is the issue?

yenchenlin replied:
I originally thought the input data is explicitly cast to float64 before the call to `_preprocess_data` for the other models, but then found that some are not.

I am afraid of some models assume float32 data will be transformed into float64 after `_preprocess_data`.

MechCoder replied:
OK, I still do not get why that will be an issue unless the other models use Cython, but I suppose this is ok as a temporary solution.

yenchenlin committed:
Fix error msg

yenchenlin committed:
Move declarartion

yenchenlin committed:
Remove redundant comment

yenchenlin committed:
Add tests

yenchenlin commented:
Comments addressed!

MechCoder commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -670,13 +675,13 @@ def fit(self, X, y, check_input=True):
         # We expect X and y to be already float64 Fortran ordered arrays
```
Not your fault, but you should change this comment about float64 input.

MechCoder replied:
Did you remove this yet? It does not show in the diff view

MechCoder commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -664,3 +664,31 @@ def test_lasso_non_float_y():
+    # Here we have a small number of iterations, and thus the
+    # ElasticNet might not converge. This is to speed up tests
+
+    for fit_intercept in [True, False]:
```
Seems like we should also check with `normalize in [True, False]`

MechCoder commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -470,7 +473,9 @@ def enet_path(X, y, l1_ratio=0.5, eps=1e-3, n_alphas=100, alphas=None,
                           ' You might want' +
-                          ' to increase the number of iterations',
+                          ' to increase the number of iterations.' +
+                          ' Fitting float32 data with small alpha,' +
```
This warning should definitely be decoupled with the existing warning.

You can move the other warning to the start of fit, when the dtype of X is float32 and alpha is less than 1e-8 instead of confusing users with a unrelated message when the algorithm has not converged.

yenchenlin replied:
I see, nice solution!

MechCoder commented:
@yenchenlin That is definitely my last pass! The major concern is with setting `X_scale` to the appropriate dtype to do away with the workaround. Also can you update the title of the pull request with the current status of the memory benchmarks?

jnothman commented:
Is the alpha < 1e-8 not dependent on other things, like feature scale or perhaps even number of features? I've not thought about it a great deal...

yenchenlin committed:
Test normalize

yenchenlin commented:
@jnothman I agree. It should depend both on feature scale and number of features.

yenchenlin commented:
And @MechCoder , the memory profiling results retain the same!

MechCoder commented:
LGTM. Will let @jnothman make the final pass(es)

yenchenlin commented:
Sorry @MechCoder I'm still confusing about the 1e-8 issues ...

As @jnothman mentions, it depends on the feature scales and the number of features.

MechCoder commented:
That's fine. Let us just remove that warning. It is highly unlikely that a user will set such a low value of alpha (10**-8) anyway

yenchenlin committed:
Delete warning

yenchenlin commented:
I'm thinking of leaving a convergence warning such as

"Fitting float32 data with a really low alpha value may cause precision issues."

I think it's better than nothing.

yenchenlin commented:
Actually, [parameter doc of `alpha`](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.ElasticNet.html) has mentioned it ...


> alpha = 0 is equivalent to an ordinary least square, solved by the LinearRegression object. For numerical reasons, using alpha = 0 with the Lasso object is not advised and you should prefer the LinearRegression object.

yenchenlin commented:
How about a convergence warning:

 "Fitting float32 data with a alpha near 0 may cause numerical issues, consider using LinearRegression instead."

MechCoder commented:
That comment should be valid irrespective of whether X if loaf32 or float64 no?

yenchenlin commented:
Oh yes, so should we add it?

MechCoder commented:
Sure, but that will not influence my +1 in any way.

yenchenlin commented:
ooops, but adding the warning still requires me to determine a concrete value, which I can't.

I'll wait for @jnothman 's opinion.

jnothman commented:
> ooops, but adding the warning still requires me to determine a concrete value, which I can't.

Once we know it has not converged, adding a warning for alpha < 1e-8 or similar might be sensible (or some constant identified on normalized data; are you sure number of features matters?)

yenchenlin committed:
Fix comment

yenchenlin committed:
Add error msg

yenchenlin commented:
I've added back the error message that informs user to increase `alpha` once they encounter non-convergence error.

I think all comments are addressed now.

jnothman commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -473,7 +473,9 @@ def enet_path(X, y, l1_ratio=0.5, eps=1e-3, n_alphas=100, alphas=None,
                           ' You might want' +
-                          ' to increase the number of iterations.',
+                          ' to increase the number of iterations.' +
+                          ' Fitting data with alpha near zero, e.g., 1e-8,' +
```
*Shrugs*

I suppose this will do. I think mentioning float32 is better, but still...

yenchenlin replied:
okay 🎅  I add float32.

jnothman commented:
LGTM. Please add this to what's new (either as its own entry or together with related fused type fixes).

yenchenlin committed:
Add error msg

yenchenlin committed:
Add what's new

yenchenlin commented:
@jnothman done!

jnothman commented:
Have you forgotten to push those last changes? whats_new does not appear updated, nor the warning.

jnothman commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -474,7 +474,8 @@ def enet_path(X, y, l1_ratio=0.5, eps=1e-3, n_alphas=100, alphas=None,
                           ' You might want' +
                           ' to increase the number of iterations.' +
-                          ' Fitting data with alpha near zero, e.g., 1e-8,' +
+                          ' Fitting float32 data with alpha near zero,' +
```
But this is only relevant if the data is float32, no?

yenchenlin replied:
Actually I think fitting with a really small alpha, even float64 data may not converge.

jnothman replied:
Sure, so make the warning as relevant and useful as possible to a user that triggers it.

yenchenlin replied:
So is simply remove the **float32** enough 😛?

jnothman replied:
Not really, because alpha=1e-8 isn't ordinarily too small for normalized float64. Either remove reference to the alpha value and you're giving a much more meaningful message. Also "alpha near zero" would usually be "very small alpha".

yenchenlin replied:
I have to admit that I'm not very sure about **all** the factors that will cause convergence issue, and thus not dare to determine a specific reference value.

Or we can remove reference to the alpha value?

jnothman replied:
Remove any specific value. Just say near zero

On 25 August 2016 at 23:34, Yen <notifications@github.com> wrote:

> In sklearn/linear_model/coordinate_descent.py
> <https://github.com/scikit-learn/scikit-learn/pull/6913#discussion_r76242986>
> :
>
> > @@ -474,7 +474,8 @@ def enet_path(X, y, l1_ratio=0.5, eps=1e-3, n_alphas=100, alphas=None,
> >              warnings.warn('Objective did not converge.' +
> >                            ' You might want' +
> >                            ' to increase the number of iterations.' +
> > -                          ' Fitting data with alpha near zero, e.g., 1e-8,' +
> > +                          ' Fitting float32 data with alpha near zero,' +
>
> I have to admit that I'm not very sure about *all* the factors that will
> cause convergence issue, and thus not dare to determine a specific
> reference value.
>
> Or we can remove reference to the alpha value?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6913/files/82fdf0962e3c7b0965b54ca137a56ab6d01fc226..c032d3b5820b53fb0717008435a44245cdb746f1#r76242986>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz65p1hO8jA-O7s1Ka5BIexYN_p7qlks5qjZnKgaJpZM4I6SId>
> .
>

yenchenlin replied:
done!

MechCoder commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -470,7 +473,10 @@ def enet_path(X, y, l1_ratio=0.5, eps=1e-3, n_alphas=100, alphas=None,
                           ' You might want' +
-                          ' to increase the number of iterations',
+                          ' to increase the number of iterations.' +
+                          ' Fitting data with alpha near zero,' +
```
Did we not agree to add this message only when `alpha` is less than some heuristic value?

yenchenlin replied:
It seems hard to determine the *heuristic* value 😭

MechCoder commented:
Your benchmarks in the description of the Pull Request also suggests non-trivial speed gains. Do the speed gains also still hold?

yenchenlin commented:
@MechCoder yes!

- float32
![32](https://cloud.githubusercontent.com/assets/7057863/17970896/06e78d12-6b0b-11e6-9679-b524a42dc972.png)

- float64
![64](https://cloud.githubusercontent.com/assets/7057863/17970897/06ee0ae8-6b0b-11e6-96a4-993fa32ee418.png)

yenchenlin committed:
Fix error msg

MechCoder commented:
Awesome! Merging with master and thanks a lot for your perseverance.

yenchenlin commented:
😭🍻  😭🍻  😭🍻

MechCoder commented:
It should be worth adding a note to `src/cblas/README.txt` to let know what changes have to be made to add to call cblas functions internally. Maybe @fabianp can do that?

jnothman commented:
Hurrah! Thanks @fabianp for rescuing this. And to @MechCoder for inviting that saviour. And to @yenchenlin for winning.

MechCoder commented:
And you are the dark knight? :p

jnothman commented:
Just a heads up that I'm a little concerned that these changes to Lasso changed its behaviour (for float64 data). It seems to have resulted in a test failure at https://github.com/scikit-learn/scikit-learn/pull/6717#issuecomment-244842499. For the dummy data I've tried so far, behaviour isn't changed, so this needs more verification.
